[explore] proper filtering of NULLs and ''

mistercrunch opened the pull request:
<img width="540" alt="screen shot 2018-03-19 at 11 21 00 pm" src="https://user-images.githubusercontent.com/487433/37638977-ce76704a-2bcc-11e8-9e80-f717607c3a0e.png">

closes https://github.com/apache/incubator-superset/issues/4297

mistercrunch committed:
[WiP] [explore] proper filtering of NULLs and ''

TODO: handling of Druid equivalents

mistercrunch committed:
Unit tests

hughhhh commented:
Definitely need this!!

hughhhh commented:
Did some work on the druid side. 

https://github.com/hughhhh/incubator-superset/commit/3bc3bd46d05a96335f9663745f68429f10b5eff3

jeffreythewang commented on superset/assets/javascripts/components/AlteredSliceTag.jsx:
```diff
@@ -61,7 +61,7 @@ export default class AlteredSliceTag extends React.Component {
       }
       return value.map((v) => {
-        const filterVal = v.val.constructor === Array ? `[${v.val.join(', ')}]` : v.val;
+        const filterVal = v.val && v.val.constructor === Array ? `[${v.val.join(', ')}]` : v.val;
```
might be my unfamiliarity, but why not use `Array.isArray` like in other parts of the code?

mistercrunch replied:
Didn't write that line but FWIW someone on StackOverflow says it's the best way
https://stackoverflow.com/questions/767486/how-do-you-check-if-a-variable-is-an-array-in-javascript

jeffreythewang reviewed (commented):
I had a [WIP] idea (https://github.com/tc-dc/superset/pull/8) for handling nulls that involves showing a tooltip next to the label, to distinguish between actual `null` values and `"null"` string values.

(I've only tested with druid, and so far only have it such that `Enable Filter Select` must be on.)

mistercrunch committed:
Some refactoring

mistercrunch committed:
[druid] fix 'Unorderable types' when col has nuls

Error "unorderable types: str() < int()" occurs when grouping by a
numerical Druid colummn that contains null values.

* druid/pydruid returns strings in the datafram with NAs for nulls
* Superset has custom logic around get_fillna_for_col that fills in the
NULLs based on declared column type (FLOAT here), so now we have a mixed
bag of type in the series
* pandas chokes on pivot_table or groupby operations as it cannot sorts
mixed types

The approach here is to stringify and fillna('<NULL>') to get a
consistent series.

mistercrunch committed:
typo

mistercrunch committed:
Fix druid_func tests

betodealmeida commented on superset/connectors/base/models.py:
```diff
@@ -185,6 +186,34 @@ def data(self):
+        def handle_single_value(v):
+            # backward compatibility with previous <select> components
+            if isinstance(v, basestring):
+                v = v.strip().strip("'").strip('"')
```
You can do a single strip to remove whitespace and single/double quotes:

```python
v = v.strip(' \'"')
```

Or, to also take care of tabs and line feeds:

```python
v = v.strip('\t\n \'"')
```

mistercrunch replied:
oh neat

betodealmeida commented on superset/connectors/base/models.py:
```diff
@@ -185,6 +186,34 @@ def data(self):
+            return v
+        if isinstance(values, (list, tuple)):
+            values = [handle_single_value(v) for v in values]
+        values = handle_single_value(values)
```
This is a bit confusing... I'd add the second call to an `else` block (unless I'm missing something):

```python
if isinstance(values, (list, tuple)):
    values = [handle_single_value(v) for v in values]
else:
    values = handle_single_value(values)
```

betodealmeida commented on superset/connectors/base/models.py:
```diff
@@ -185,6 +186,34 @@ def data(self):
+            values = [handle_single_value(v) for v in values]
+        values = handle_single_value(values)
+        if is_list_target and not isinstance(values, (tuple, list)):
+            values = [values]
```
What if `is_list_target` is true but `values` is a tuple? In this case it would return a tuple, is that ok?

mistercrunch replied:
The variable should probably be called `is_iterable_target`, both tuple and list will work here.

betodealmeida commented on superset/connectors/druid/models.py:
```diff
@@ -1322,40 +1339,30 @@ def increment_timestamp(ts):
-                    eq = utils.string_to_num(eq)
-
+            eq = cls.filter_values_handler(
+                eq, is_list_target=op in ('in', 'not in'),
```
This is a bit hard to read, can we define `is_list_target` outside the function call, like you did for `is_numeric_col`?

betodealmeida commented on superset/connectors/sqla/models.py:
```diff
@@ -565,28 +564,20 @@ def get_sqla_query(  # sqla
+            eq = self.filter_values_handler(
+                flt.get('val'),
+                target_column_is_numeric=col_obj.is_num,
+                is_list_target=op in ('in', 'not in'))
```
Same as above here.

betodealmeida commented on superset/viz.py:
```diff
@@ -1693,6 +1693,8 @@ def run_extra_queries(self):
             qry['groupby'] = [flt]
             df = self.get_df_payload(query_obj=qry).get('df')
+            print(df)
+            print(df.dtypes)
```
`d2d`

betodealmeida commented on tests/druid_func_tests.py:
```diff
@@ -119,14 +119,15 @@ def test_get_filters_handles_arrays_for_string_types(self):
         filtr = {'col': 'A', 'op': '==', 'val': []}
         res = DruidDatasource.get_filters([filtr], [])
-        self.assertEqual('', res.filter['filter']['value'])
+        self.assertEqual(None, res.filter['filter']['value'])
```
Use `assertIsNone` instead.

betodealmeida commented on tests/druid_func_tests.py:
```diff
@@ -119,14 +119,15 @@ def test_get_filters_handles_arrays_for_string_types(self):
         filtr = {'col': 'A', 'op': '==', 'val': None}
         res = DruidDatasource.get_filters([filtr], [])
-        self.assertEqual('', res.filter['filter']['value'])
+        self.assertEqual(None, res)
```
Ditto.

mistercrunch committed:
Addressing more comments

mistercrunch committed:
last touches
